File: neo/frontend/lib/api/embedded-sites.ts
/**
 * Embedded Sites API Client
 * Handles all embedded site related API calls
 */

import type {
	EmbeddedSite,
	CreateEmbeddedSiteInput,
	UpdateEmbeddedSiteInput,
} from "@/components/embedded-site";
import type { ApiResponse } from "@/components/workspace/workspace-types";

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL || "";

// ============================================================
// API Fetch Wrapper
// ============================================================
async function apiFetch<T>(
	endpoint: string,
	options: RequestInit = {},
): Promise<ApiResponse<T>> {
	const rawToken =
		typeof window !== "undefined" ? localStorage.getItem("neo-auth") : null;
	const parsed = rawToken ? JSON.parse(rawToken) : null;
	const token = parsed?.state?.user?.token ?? null;

	const response = await fetch(`${API_BASE_URL}${endpoint}`, {
		...options,
		headers: {
			"Content-Type": "application/json",
			...(token && { Authorization: `Bearer ${token}` }),
			...options.headers,
		},
		credentials: "include",
	});

	const data = await response.json();

	if (!response.ok) {
		throw {
			code: data.code || response.status,
			message: data.message || "请求失败",
			detail: data.detail,
		};
	}

	// API may return data without code field (success by default) or with code=0
	// Only throw error if code exists and is non-zero
	if (data.code !== undefined && data.code !== 0) {
		throw {
			code: data.code,
			message: data.message || "请求失败",
			detail: data.detail,
		};
	}

	return data as ApiResponse<T>;
}

// ============================================================
// Embedded Sites API
// ============================================================

export interface EmbeddedSiteListResponse {
	items: EmbeddedSite[];
	total: number;
	page: number;
	page_size: number;
	total_pages: number;
}

export interface EmbeddedSiteStatusResponse {
	id: number;
	status: string;
	updated_at: string;
}

export interface ListEmbeddedSitesParams {
	page?: number;
	page_size?: number;
	status?: "ENABLED" | "DISABLED";
	search?: string;
}

/**
 * List embedded sites for a workspace
 * GET /api/v1/workspaces/{workspace_code}/embedded-sites
 */
export async function listEmbeddedSites(
	workspaceCode: string,
	params: ListEmbeddedSitesParams = {},
): Promise<{
	list: EmbeddedSite[];
	total: number;
	page: number;
	page_size: number;
	total_pages: number;
}> {
	const searchParams = new URLSearchParams();
	if (params.page) searchParams.set("page", String(params.page));
	if (params.page_size) searchParams.set("page_size", String(params.page_size));
	if (params.status) searchParams.set("status", params.status);
	if (params.search) searchParams.set("search", params.search);

	const query = searchParams.toString();
	const response = await apiFetch<EmbeddedSiteListResponse>(
		`/api/v1/workspaces/${workspaceCode}/embedded-sites${query ? `?${query}` : ""}`,
	);

	return {
		list: response.data?.items || [],
		total: response.data?.total || 0,
		page: response.data?.page || 1,
		page_size: response.data?.page_size || 20,
		total_pages: response.data?.total_pages || 0,
	};
}

/**
 * Get single embedded site
 * GET /api/v1/workspaces/{workspace_code}/embedded-sites/{id}
 */
export async function getEmbeddedSite(
	workspaceCode: string,
	siteId: number,
): Promise<EmbeddedSite> {
	const response = await apiFetch<EmbeddedSite>(
		`/api/v1/workspaces/${workspaceCode}/embedded-sites/${siteId}`,
	);
	return response.data;
}

/**
 * Create embedded site
 * POST /api/v1/workspaces/{workspace_code}/embedded-sites
 */
export async function createEmbeddedSite(
	workspaceCode: string,
	data: CreateEmbeddedSiteInput,
): Promise<EmbeddedSite> {
	const response = await apiFetch<EmbeddedSite>(
		`/api/v1/workspaces/${workspaceCode}/embedded-sites`,
		{
			method: "POST",
			body: JSON.stringify(data),
		},
	);
	return response.data;
}

/**
 * Update embedded site
 * PUT /api/v1/workspaces/{workspace_code}/embedded-sites/{id}
 */
export async function updateEmbeddedSite(
	workspaceCode: string,
	siteId: number,
	data: UpdateEmbeddedSiteInput,
): Promise<EmbeddedSite> {
	const response = await apiFetch<EmbeddedSite>(
		`/api/v1/workspaces/${workspaceCode}/embedded-sites/${siteId}`,
		{
			method: "PUT",
			body: JSON.stringify(data),
		},
	);
	return response.data;
}

/**
 * Delete embedded site (soft delete)
 * DELETE /api/v1/workspaces/{workspace_code}/embedded-sites/{id}
 */
export async function deleteEmbeddedSite(
	workspaceCode: string,
	siteId: number,
): Promise<void> {
	await apiFetch<void>(
		`/api/v1/workspaces/${workspaceCode}/embedded-sites/${siteId}`,
		{
			method: "DELETE",
		},
	);
}

/**
 * Enable embedded site
 * PATCH /api/v1/workspaces/{workspace_code}/embedded-sites/{id}/enable
 */
export async function enableEmbeddedSite(
	workspaceCode: string,
	siteId: number,
): Promise<EmbeddedSiteStatusResponse> {
	const response = await apiFetch<EmbeddedSiteStatusResponse>(
		`/api/v1/workspaces/${workspaceCode}/embedded-sites/${siteId}/enable`,
		{
			method: "PATCH",
		},
	);
	return response.data;
}

/**
 * Disable embedded site
 * PATCH /api/v1/workspaces/{workspace_code}/embedded-sites/{id}/disable
 */
export async function disableEmbeddedSite(
	workspaceCode: string,
	siteId: number,
): Promise<EmbeddedSiteStatusResponse> {
	const response = await apiFetch<EmbeddedSiteStatusResponse>(
		`/api/v1/workspaces/${workspaceCode}/embedded-sites/${siteId}/disable`,
		{
			method: "PATCH",
		},
	);
	return response.data;
}

// ============================================================
// Error Helper
// ============================================================
export function getErrorMessage(code: number): string {
	const messages: Record<number, string> = {
		0: "成功",
		1001: "参数错误",
		1002: "未授权",
		1003: "禁止访问",
		2001: "嵌入网站不存在",
		2002: "嵌入网站名称已存在",
		2003: "存在关联的 Agent",
		3001: "工作区不存在",
		3002: "工作区名称已存在",
		9001: "服务器内部错误",
	};
	return messages[code] || "请求失败，请稍后重试";
}

File: neo/backend/src/app/api/v1/embedded_sites.py
"""API routes for embedded sites."""

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies import get_current_user
from app.models.user import User
from app.repositories import workspace_repository
from app.repositories.embedded_site_repository import EmbeddedSiteRepository
from app.schemas.embedded_site import (
    EmbeddedSiteCreate,
    EmbeddedSiteListResponse,
    EmbeddedSiteResponse,
    EmbeddedSiteStatusResponse,
    EmbeddedSiteUpdate,
)
from app.schemas.response import ApiResponse

router = APIRouter(prefix="/workspaces/{workspace_code}/embedded-sites", tags=["embedded-sites"])


def _get_workspace_id(
    workspace_code: str,
    db: Session,
) -> int:
    """Get workspace ID by code, raise 404 if not found."""
    workspace = workspace_repository.get_workspace_by_code(db, workspace_code)

    if not workspace:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workspace not found",
        )

    # TODO: Check workspace membership
    return int(workspace.id)


@router.get("", response_model=ApiResponse[EmbeddedSiteListResponse])
def list_embedded_sites(
    workspace_code: str,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    site_status: str | None = Query(None, alias="status"),
    search: str | None = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> EmbeddedSiteListResponse:
    """List embedded sites for a workspace with pagination, filtering, and search."""
    workspace_id = _get_workspace_id(workspace_code, db)
    repo = EmbeddedSiteRepository(db)

    sites, total = repo.list_by_workspace(
        workspace_id=workspace_id,
        page=page,
        page_size=page_size,
        status=site_status,
        search=search,
    )

    total_pages = (total + page_size - 1) // page_size if total > 0 else 0

    return ApiResponse.success(
        data=EmbeddedSiteListResponse(
            items=[EmbeddedSiteResponse.model_validate(site) for site in sites],
            total=total,
            page=page,
            page_size=page_size,
            total_pages=total_pages,
        ),
    )


@router.get("/{site_id}", response_model=ApiResponse[EmbeddedSiteResponse])
def get_embedded_site(
    workspace_code: str,
    site_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ApiResponse[EmbeddedSiteResponse]:
    """Get a single embedded site by ID."""
    _get_workspace_id(workspace_code, db)
    repo = EmbeddedSiteRepository(db)

    site = repo.get_by_id(site_id)
    if not site:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Embedded site not found",
        )

    return ApiResponse.success(data=EmbeddedSiteResponse.model_validate(site))


@router.post("", response_model=ApiResponse[EmbeddedSiteResponse], status_code=status.HTTP_201_CREATED)
def create_embedded_site(
    workspace_code: str,
    data: EmbeddedSiteCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ApiResponse[EmbeddedSiteResponse]:
    """Create a new embedded site."""
    workspace_id = _get_workspace_id(workspace_code, db)
    repo = EmbeddedSiteRepository(db)

    # Check for duplicate name within workspace
    existing = repo.get_by_workspace_and_name(workspace_id, data.site_name)
    if existing:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Site name already exists in workspace",
        )

    site = repo.create(workspace_id=workspace_id, user_id=int(current_user.id), data=data)
    return ApiResponse.success(data=EmbeddedSiteResponse.model_validate(site))


@router.put("/{site_id}", response_model=ApiResponse[EmbeddedSiteResponse])
def update_embedded_site(
    workspace_code: str,
    site_id: int,
    data: EmbeddedSiteUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ApiResponse[EmbeddedSiteResponse]:
    """Update an existing embedded site."""
    _get_workspace_id(workspace_code, db)
    repo = EmbeddedSiteRepository(db)

    site = repo.get_by_id(site_id)
    if not site:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Embedded site not found",
        )

    # Check for duplicate name if site_name is being updated
    if data.site_name and data.site_name != site.site_name:
        existing = repo.get_by_workspace_and_name(int(site.workspace_id), data.site_name, exclude_id=site_id)
        if existing:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Site name already exists in workspace",
            )

    updated_site = repo.update(site, data)
    return ApiResponse.success(data=EmbeddedSiteResponse.model_validate(updated_site))


@router.delete("/{site_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_embedded_site(
    workspace_code: str,
    site_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> None:
    """Delete an embedded site (soft delete)."""
    _get_workspace_id(workspace_code, db)
    repo = EmbeddedSiteRepository(db)

    site = repo.get_by_id(site_id)
    if not site:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Embedded site not found",
        )

    # TODO: Check for linked agents before deletion
    # This check will be implemented when agent-embedded-site association is added

    repo.soft_delete(site)


@router.patch("/{site_id}/enable", response_model=ApiResponse[EmbeddedSiteStatusResponse])
def enable_embedded_site(
    workspace_code: str,
    site_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ApiResponse[EmbeddedSiteStatusResponse]:
    """Enable a disabled embedded site."""
    _get_workspace_id(workspace_code, db)
    repo = EmbeddedSiteRepository(db)

    site = repo.get_by_id(site_id)
    if not site:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Embedded site not found",
        )

    enabled_site = repo.enable(site)
    return ApiResponse.success(
        data=EmbeddedSiteStatusResponse(
            id=int(enabled_site.id),
            status=enabled_site.status.value,
            updated_at=enabled_site.updated_at,
        ),
    )


@router.patch("/{site_id}/disable", response_model=ApiResponse[EmbeddedSiteStatusResponse])
def disable_embedded_site(
    workspace_code: str,
    site_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ApiResponse[EmbeddedSiteStatusResponse]:
    """Disable an enabled embedded site."""
    _get_workspace_id(workspace_code, db)
    repo = EmbeddedSiteRepository(db)

    site = repo.get_by_id(site_id)
    if not site:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Embedded site not found",
        )

    disabled_site = repo.disable(site)
    return ApiResponse.success(
        data=EmbeddedSiteStatusResponse(
            id=int(disabled_site.id),
            status=disabled_site.status.value,
            updated_at=disabled_site.updated_at,
        ),
    )
